### PR TITLE
Improvement of pool allocator

### PR DESCRIPTION
**Pull request by ruben-ayrapetyan**
Arndale (Dual-core, 1 Ghz)

Benchmark | Rss | Perf
-------- | -------- | --------
./sunspider-1.0.2.stable/3d-cube.js | 176 -> 172 <br> (2.273%) | 3.536 -> 3.03 <br> (14.310%) | 
./sunspider-1.0.2.stable/access-binary-trees.js | 100 -> 104 <br> (-4.000%) | 1.993 -> 2.124 <br> (-6.573%) | 
./sunspider-1.0.2.stable/access-fannkuch.js | 48 -> 48 <br> (0.000%) | 8.566 -> 7.237 <br> (15.515%) | 
./sunspider-1.0.2.stable/access-nbody.js | 72 -> 72 <br> (0.000%) | 3.908 -> 3.488 <br> (10.747%) | 
./sunspider-1.0.2.stable/bitops-3bit-bits-in-byte.js | 36 -> 36 <br> (0.000%) | 2.629 -> 2.328 <br> (11.449%) | 
./sunspider-1.0.2.stable/bitops-bits-in-byte.js | 36 -> 36 <br> (0.000%) | 3.722 -> 3.181 <br> (14.535%) | 
./sunspider-1.0.2.stable/bitops-bitwise-and.js | 28 -> 28 <br> (0.000%) | 2.269 -> 1.881 <br> (17.100%) | 
./sunspider-1.0.2.stable/controlflow-recursive.js | 216 -> 224 <br> (-3.704%) | 2.084 -> 1.882 <br> (9.693%) | 
./sunspider-1.0.2.stable/crypto-sha1.js | 148 -> 168 <br> (-13.514%) | 64.046 -> 56.812 <br> (11.295%) | 
./sunspider-1.0.2.stable/date-format-xparb.js | 112 -> 112 <br> (0.000%) | 2.321 -> 2.348 <br> (-1.163%) | 
./sunspider-1.0.2.stable/math-cordic.js | 52 -> 48 <br> (7.692%) | 4.637 -> 3.751 <br> (19.107%) | 
./sunspider-1.0.2.stable/math-partial-sums.js | 40 -> 40 <br> (0.000%) | 1.819 -> 1.647 <br> (9.456%) | 
./sunspider-1.0.2.stable/math-spectral-norm.js | 52 -> 52 <br> (0.000%) | 2.557 -> 2.068 <br> (19.124%) | 
./sunspider-1.0.2.stable/string-fasta.js | 60 -> 52 <br> (13.333%) | 19.614 -> 19.52 <br> (0.479%) | 
Geometric mean | 0.3122% | 10.659%

Benchmark | Rss | Perf
-------- | -------- | --------
./sunspider-0.9.1/3d-cube.js | 9.302% | 14.608% | 
./sunspider-0.9.1/access-binary-trees.js | -4.348% | 10.991% | 
./sunspider-0.9.1/access-fannkuch.js | 15.385% | 9.101% | 
./sunspider-0.9.1/access-nbody.js | 5.556% | 10.877% | 
./sunspider-0.9.1/bitops-3bit-bits-in-byte.js | 11.111% | 12.732% | 
./sunspider-0.9.1/bitops-bits-in-byte.js | 0.000% | 14.623% | 
./sunspider-0.9.1/bitops-bitwise-and.js | 0.000% | 6.733% | 
./sunspider-0.9.1/controlflow-recursive.js | -5.660% | 9.434% | 
./sunspider-0.9.1/date-format-xparb.js | 0.000% | -1.037% | 
./sunspider-0.9.1/math-cordic.js | -9.091% | 17.247% | 
./sunspider-0.9.1/math-partial-sums.js | 0.000% | 12.784% | 
./sunspider-0.9.1/math-spectral-norm.js | 0.000% | 19.504% | 
Geometric mean | 2.1026% | 11.6084%

### CODE REVIEW

File: jerry-core/mem/mem-poolman.cpp [hunk 1]
```diff
@@ -86,61 +165,325 @@ mem_pools_init (void)
...
+
+  /*
+   * At first pass collect pointers to those of free chunks that are first at their pools
+   * to separate lists (collection-time pool lists) and change them to headers of corresponding pools
+   */
+
+  /*
+   * Number of collection-time pool lists
+   */
+  constexpr uint32_t pool_lists_number = 8;
```
Comment (egavrin) [inline jerry-core/mem/mem-poolman.cpp]: Why `constrexpr`?

File: jerry-core/mem/mem-poolman.cpp [hunk 2]
```diff
@@ -86,61 +165,325 @@ mem_pools_init (void)
...
-  JERRY_ASSERT (mem_pools == NULL);
+  mem_pools_collect_empty_pools ();
+
+#ifndef JERRY_NDEBUG
   JERRY_ASSERT (mem_free_chunks_number == 0);
+#endif /* !JERRY_NDEBUG */
 } /* mem_pools_finalize */
 
+void
+mem_pools_collect_empty_pools (void)
```
Comment (egavrin) [inline jerry-core/mem/mem-poolman.cpp]: `mem_pools_collect_empty`?

File: jerry-core/mem/mem-poolman.cpp [hunk 3]
```diff
@@ -86,61 +165,325 @@ mem_pools_init (void)
...
     }
 
-    JERRY_ASSERT (prev_pool_state_p != NULL && pool_state != mem_pools);
+    prev_free_chunk_p = chunk_p;
+  }
+
+  prev_free_chunk_p->u.free.next_p = NULL;
+
+#ifdef JERRY_VALGRIND
+  for (mem_pool_chunk_index_t chunk_index = 0;
```
Comment (egavrin) [inline jerry-core/mem/mem-poolman.cpp]: `for (mem_pool_chunk_index_t chunk_index = 0; chunk_index < MEM_POOL_CHUNKS_NUMBER; chunk_index++)`

File: jerry-core/mem/mem-poolman.cpp (hunk 2 above)
Comment (ruben-ayrapetyan) [inline jerry-core/mem/mem-poolman.cpp]: Ok. Thanks!

File: jerry-core/mem/mem-poolman.cpp (hunk 1 above)
Comment (ruben-ayrapetyan) [inline jerry-core/mem/mem-poolman.cpp]: To make possible declaration of constant-sized array below.

File: jerry-core/mem/mem-poolman.cpp (hunk 3 above)
Comment (ruben-ayrapetyan) [inline jerry-core/mem/mem-poolman.cpp]: Seems to be too long for convenient reading.

File: jerry-core/mem/mem-poolman.cpp [hunk 4]
```diff
@@ -86,61 +165,325 @@ mem_pools_init (void)
...
+#endif /* !JERRY_NDEBUG */
 } /* mem_pools_finalize */
 
+void
+mem_pools_collect_empty (void)
+{
+  /*
+   * Hint magic number in header of pools with free first chunks
+   */
+  const uint16_t hint_magic_num_value = 0x7e89;
```
Comment (zherczeg) [inline jerry-core/mem/mem-poolman.cpp]: Why did you choose this number?

Comment (ruben-ayrapetyan) [inline jerry-core/mem/mem-poolman.cpp]: Its a random number.

File: jerry-core/mem/mem-poolman.cpp [hunk 5]
```diff
@@ -86,61 +165,325 @@ mem_pools_init (void)
...
+   */
+  for (mem_pool_chunk_t *free_chunk_iter_p = mem_free_chunk_p, *next_free_chunk_p;
+       free_chunk_iter_p != NULL;
+       free_chunk_iter_p = next_free_chunk_p)
+  {
+    next_free_chunk_p = free_chunk_iter_p->u.free.next_p;
+
+    VALGRIND_NOACCESS_SPACE (free_chunk_iter_p, MEM_POOL_CHUNK_SIZE);
+  }
+} /* mem_pools_collect_empty */
```
Comment (zherczeg) [inline jerry-core/mem/mem-poolman.cpp]: Can we put a preprocessor guard around this code?

File: tests/unit/test-poolman.cpp [hunk 6]
```diff
@@ -20,16 +20,15 @@
...
 
 #include "test-common.h"
 
 // Iterations count
-const uint32_t test_iters = 16384;
+const uint32_t test_iters = 1024;
 
 // Subiterations count
-const uint32_t test_max_sub_iters = 32;
+const uint32_t test_max_sub_iters = 1024;
```
Comment (zherczeg) [inline tests/unit/test-poolman.cpp]: 1024*1024 is twice as big as 16384 * 32. Is this ok?

File: jerry-core/mem/mem-poolman.cpp (hunk 5 above)
Comment (ruben-ayrapetyan) [inline jerry-core/mem/mem-poolman.cpp]: Yes, of course.

**Comment by zherczeg** [comment]
LGTM but there are some questions above.

File: jerry-core/mem/mem-poolman.cpp (hunk 5 above)
Comment (ruben-ayrapetyan) [inline jerry-core/mem/mem-poolman.cpp]: Done.

File: jerry-core/mem/mem-poolman.cpp (hunk 4 above)
Comment (zherczeg) [inline jerry-core/mem/mem-poolman.cpp]: Ok, but certainly there has to be some decision behind it. Why not 0 for example? Is there a better number?

File: tests/unit/test-poolman.cpp (hunk 6 above)
Comment (ruben-ayrapetyan) [inline tests/unit/test-poolman.cpp]: Yes, it is ok. We change sub-iterations number from 32 to 1024 to increase testing coverage.

File: jerry-core/mem/mem-poolman.cpp (hunk 4 above)
Comment (ruben-ayrapetyan) [inline jerry-core/mem/mem-poolman.cpp]: @zherczeg, I've just used `cat /dev/random | hexdump -C` and took first number from the output.
`0` can appear more frequently as it is generally used for initialization of memory locations, so just random number would be more suitable in the case.

**Comment by egavrin** [comment]
`make push`

**Pull request merged**